examples/fabric/tensor_parallel/train.py

```diff
@@ -0,0 +1,77 @@
+import lightning as L
+import torch
+import torch.nn.functional as F
+from data import RandomTokenDataset
+from lightning.fabric.strategies import ModelParallelStrategy
+from model import ModelArgs, Transformer, parallelize
+from torch.distributed.tensor.parallel import loss_parallel
+from torch.utils.data import DataLoader
+
+
+def train():
+    strategy = ModelParallelStrategy(
+        # User-defined function that applies the desired parallelizations specific to the model
+        # (TP, FSDP2, activation checkpointing, ...)
+        parallelize_fn=parallelize,
+        # Define the size of the 2D parallelism
+        # Set to "auto" to apply TP intra-node and DP inter-node
+        data_parallel_size=2,
+        tensor_parallel_size=2,
+    )
+
+    fabric = L.Fabric(accelerator="cuda", devices=4, strategy=strategy)
+    fabric.launch()
+
+    # Initialize the model
+    model_args = ModelArgs(vocab_size=32000)
+    with fabric.init_module(empty_init=True):
+        model = Transformer(model_args)
+
+    fabric.print(f"Number of model parameters: {sum(p.numel() for p in model.parameters()) / 1e9:.1f} B")
+
+    # Define the optimizer
+    optimizer = torch.optim.AdamW(model.parameters(), lr=3e-3, foreach=True)
```

Comment:
With these new Pytorch abstractions, isn't it still a good idea to init the optimizer after fabric.setup?

Reply:
This requirement has been lifted a while ago in FSDP with `use_orig_parameters` and in FSDP2 it doesn't exist anymore. So I would say no?